gui/default/syncthing/folder/ignoreTreeService.js

```diff
@@ -0,0 +1,100 @@
+angular.module('syncthing.folder')
+    .service('IgnoreTree', function (
+        Ignores,
+        Browse,
+    ) {
+        'use strict';
+
+        // Bind methods directly to the controller so we can use controllerAs in template
+        var self = this;
+
+        // public definitions
+        self.tree = null;
+
+        self.refresh = function(folderId) {
+            var promise = Browse.refresh(folderId).then(function(response) {
+                return response.files.map(buildNode);
+            });
+
+            if (self.tree) {
+                self.tree.reload(promise);
+            } else {
+                self.tree = $("#ignore-tree").fancytree({
+                    extensions: ["table"],
+                    checkbox: true,
+                    selectMode: 2,
+                    table: {
+                        indentation: 20,
+                        nodeColumnIdx: 1,
+                        checkboxColumnIdx: 0,
+                    },
+                    debugLevel: 2,
+                    source: promise,
+                    lazyLoad: function (event, data) {
+                        var prefix = data.node.data.file.path;
+                        data.result = Browse.refresh(folderId, prefix).then(function(response) {
+                            return response.files.map(buildNode);
+                        });
+                    },
+                    select: function (event, data) {
+                        toggle(data.node);
+                    },
+                }).fancytree("getTree");
+            }
+        }
+
+        self.update = function() {
+            if (self.tree) updateNodes(self.tree.rootNode.children)
+        };
+
+        function buildNode(file) {
+            var match = Ignores.matchingPattern(file);
+
+            return {
+                // Fancytree keys
+                title: file.name,
+                selected: (!match || match.isNegated),
+                key: file.path,
+                lazy: !file.isFile,
+                folder: !file.isFile,
+                // Data keys
+                file: file,
+                match: match,
+            };
+        };
+
+        function toggle(node) {
+            var absPath = '/' + node.data.file.path;
+            if (node.data.match) {
+                var match = node.data.match;
+                if (absPath === match.path) {
+                    // match is exact match to this file, remove match from patterns
+                    Ignores.removePattern(match.text);
```

Comment:
Shouldn't this check if the pattern is negated?
Sure, it might be an exact match, but it might be negated, at which point removing it would do the wrong thing?

Also, I guess "toggle" doesn't actually say whether it was enabled or disabled, so we're just assuming that we got things right from the get go?

Reply:
I guess I need to try this out to understand how it works and how it interacts with existing patterns, but this raises a few questions/points.

1. It seems that in the case above, we would add a redundant pattern, is that the case, or the example is just to explain handling of pre-existing text patterns placed there by the user?
2. We should try to "collapse" patterns as much as we can, as it's N files in a folder matched against M patterns.
3. What happens when the existing text based patterns created by a user are a mess? I guess the tree ui does not try to simplify it?

Reply:
1. It is to explain handling of pre-existing text pattern, but it's _possible_ you could end up with redundant patterns from the UI, with the right sequence of changes.
2. Collapsing would resolve that issue. I'll investigate that. The bones are there to determine when one pattern's path is the child of another so I hope it is not too complicated.
3. The UI will make no changes to the patterns until a file is toggled in the UI. Even for messy patterns, it should be representing their effects accurately though!